### PR TITLE
Feature: Weakly connected components helper functions

### PR DESCRIPTION
**Pull request by njayaram2**
JIRA: MADLIB-1101

Add several helper functions that will quickly return back various
useful stats based on the connected components learng from the
madlib.weakly_connected_components() function. Five helper functions
are added as part of this story, along with docs and updated install
check. The helper functions are:
- graph_wcc_largest_cpt(): finds largest components
- graph_wcc_histogram(): finds number of vertices in each component
- graph_wcc_vertex_check(): finds all components that have a given
pair of vertices in them.
- graph_wcc_num_cpts(): finds total number of components.
- graph_wcc_reachable_vertices(): finds all vertices reachable
within a component for a given source vertex.

All these functions are implemented to handle grouping columns too
if the WCC's output table was created with grouping_cols.

Closes #155

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/127/

File: src/ports/postgres/modules/graph/graph_utils.py_in [hunk 1]
```diff
@@ -71,6 +70,43 @@ def _grp_from_table(tbl, grp_list):
                        for i in grp_list])
 
 
+def validate_output_and_summary_tables(model_out_table, module_name,
```
Comment (iyerr3) [inline src/ports/postgres/modules/graph/graph_utils.py_in]: The docstring and errors look good. I suggest removing the exclamation (`!`) from end of error messages.

File: src/ports/postgres/modules/graph/wcc.py_in [hunk 2]
```diff
@@ -81,18 +88,24 @@ def wcc(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
...
     params_types = {'src': str, 'dest': str}
     default_args = {'src': 'src', 'dest': 'dest'}
-    edge_params = extract_keyvalue_params(edge_args, params_types, default_args)
+    edge_params = extract_keyvalue_params(
+        edge_args, params_types, default_args)
 
-    # populate default values for optional params if null
+    # populate default values for optional params if null, and prepare data
+    # to be written into the summary table (*_st variable names)
     if vertex_id is None:
```
Comment (iyerr3) [inline src/ports/postgres/modules/graph/wcc.py_in]: `if not vertex_id`?

File: src/ports/postgres/modules/graph/wcc.py_in [hunk 3]
```diff
@@ -102,7 +115,8 @@ def wcc(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
     toupdate = unique_string(desp='toupdate')
     temp_out_table = unique_string(desp='tempout')
 
-    distribution = '' if is_platform_pg() else "DISTRIBUTED BY ({0})".format(vertex_id)
+    distribution = '' if is_platform_pg(
```
Comment (iyerr3) [inline src/ports/postgres/modules/graph/wcc.py_in]: please don't do this! 
If you want multi line then 
```
      ('' if is_platform_pg() else 
       "DISTRIBUTED BY ({0})".format(vertex_id))
```

File: src/ports/postgres/modules/graph/wcc.py_in [hunk 4]
```diff
@@ -322,10 +617,51 @@ def wcc_help(schema_madlib, message, **kwargs):
...
-            """out_table     TEXT, -- Output table of weakly connected components
-            grouping_col  TEXT -- Comma separated column names to group on
-                               -- (DEFAULT = NULL, no grouping)
-            """)
+            """out_table   TEXT, -- Output table of weakly connected components
+        grouping_col  TEXT -- Comma separated column names to group on
+                           -- (DEFAULT = NULL, no grouping)
+            """) + """
+
+    Once the above function is used to obtain the out_table, it can be used to
```
Comment (iyerr3) [inline src/ports/postgres/modules/graph/wcc.py_in]: Text below looks nice.

File: src/ports/postgres/modules/utilities/validate_args.py_in [hunk 5]
```diff
@@ -344,6 +344,50 @@ def get_cols_and_types(tbl):
     return list(zip(col_names, col_types))
 # -------------------------------------------------------------------------
 
+def get_col_type(tbl, col):
```
Comment (iyerr3) [inline src/ports/postgres/modules/utilities/validate_args.py_in]: There's lots of code overlap with `get_cols_and_types`. We should avoid that redundancy. 
Also, does `get_expr_type` not work for this need?

Comment (njayaram2) [inline src/ports/postgres/modules/utilities/validate_args.py_in]: You are right, `get_expr_type` works. Removing this function.

File: src/ports/postgres/modules/graph/wcc.py_in (hunk 3 above)
Comment (njayaram2) [inline src/ports/postgres/modules/graph/wcc.py_in]: The PEP8 extension in my editor did this automatically. :-/
Will change it.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/129/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/132/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/133/

**Comment by orhankislal** [comment]
LGTM +1

**Pull request closed without merging**